### PR TITLE
feat: `@uploadthing/expo`

### PR DESCRIPTION
**Pull request by juliusmarminge**
### TODOs

- [ ] validate upload progress, I don't seem to get any progress indications on native. I've tested uploading the same file from web and it does work so seems related to RNs implementation of XHR?
- [ ] verify it works deployed? I don't have an Apple Developer Account to test this with though...

### Demo Example

https://github.com/pingdotgg/uploadthing/assets/51714798/dc449522-a367-4f15-a6b7-425225f77f7f

### CODE REVIEW

**Comment by juliusmarminge** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/pingdotgg/uploadthing/583?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* `main`
  * **PR #579** <a href="https://app.graphite.dev/github/pr/pingdotgg/uploadthing/579?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #583** <a href="https://app.graphite.dev/github/pr/pingdotgg/uploadthing/583?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/pingdotgg/uploadthing/583?utm_source=stack-comment).

File: packages/uploadthing/src/internal/handler.ts [hunk 1]
```diff
@@ -185,7 +185,7 @@ export const buildRequestHandler = <
       req.headers.get("x-uploadthing-package") ?? "unknown";
 
     const clientVersion = req.headers.get("x-uploadthing-version");
-    if (clientVersion !== UPLOADTHING_VERSION) {
+    if (clientVersion != null && clientVersion !== UPLOADTHING_VERSION) {
```
Comment (juliusmarminge) [inline packages/uploadthing/src/internal/handler.ts:188]: should be part of different PR

**Comment by k2on** [comment]
Excited about this

File: examples/minimal-expo/metro.config.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+ * This repository configured Turborepo to use this cache location as well.
+ * If you have any environment variables, you can configure Turborepo to invalidate it when needed.
+ *
+ * @see https://turbo.build/repo/docs/reference/configuration#env
+ * @param {import('expo/metro-config').MetroConfig} config
+ * @returns {import('expo/metro-config').MetroConfig}
+ */
+function withTurborepoManagedCache(config) {
+  config.cacheStores = [
+    new FileStore({ root: path.join(__dirname, "node_modules/.cache/metro") }),
```
Comment (EvanBacon) [inline examples/minimal-expo/metro.config.js:44]: Use Expo's FileStore to ensure CSS works. https://docs.expo.dev/versions/latest/config/metro/#tailwind-troubleshooting

```js
const { FileStore } = require('@expo/metro-config/file-store');
```

File: examples/minimal-expo/metro.config.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  const workspaceRoot = path.resolve(projectRoot, "../..");
+
+  // #1 - Watch all files in the monorepo
+  config.watchFolders = [workspaceRoot];
+
+  // #2 - Resolve modules within the project's `node_modules` first, then all monorepo modules
+  config.resolver.nodeModulesPaths = [
+    path.resolve(projectRoot, "node_modules"),
+    path.resolve(workspaceRoot, "node_modules"),
+  ];
```
Comment (EvanBacon) [inline examples/minimal-expo/metro.config.js:28]: expo/metro-config does this automatically.

File: examples/minimal-expo/metro.config.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
+const { getDefaultConfig } = require("expo/metro-config");
+const { FileStore } = require("metro-cache");
+const path = require("path");
+
+module.exports = withTurborepoManagedCache(
```
Comment (EvanBacon) [inline examples/minimal-expo/metro.config.js:5]: Ensure you add `config.resolver.unstable_enablePackageExports = true;` to enable package exports. They're a bit buggy in Metro, but @robhogan has been looking into them a bit recently so it may improve.

File: examples/minimal-expo/metro.config.js (hunk 3 above)
Comment (juliusmarminge) [reply examples/minimal-expo/metro.config.js:28]: it's from the docs: https://docs.expo.dev/guides/monorepos/#modify-the-metro-config

File: packages/expo/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+) => {
+  const debuggerHost = Constants.expoConfig?.hostUri;
+  let url: URL;
+  try {
+    url = new URL(
+      initOpts?.url ?? "/api/uploadthing",
+      process.env.EXPO_PUBLIC_SERVER_ORIGIN ?? `http://${debuggerHost}`,
+    );
+    console.log("resolved url", url);
+  } catch (e) {
```
Comment (juliusmarminge) [inline packages/expo/src/index.ts:34]: cc @EvanBacon - how would you recommend grabbing the absolute url?

I know Expo allows for relative fetch calls (e.g. `fetch("/api/uploadthing")` but we only work with URL objects inside the uploadthing internals so we need to parse the absolute url some way.

if a string is passed in the objects below (L40-L42), UploadThing falls back to using `window.location.host` but that's not available in Expo and will throw runtime errors, we need to pass a URL object

---

tldr; is there some easy way to grab the server url from Expo?

File: examples/minimal-expo/metro.config.js (hunk 2 above)
Comment (juliusmarminge) [reply examples/minimal-expo/metro.config.js:44]: Ehhm...

CleanShot 2024-04-16 at 20 17 49

Comment (juliusmarminge) [reply examples/minimal-expo/metro.config.js:44]: ah needed to hoist the deps... classic pnpm x expo limitation

File: examples/minimal-expo/metro.config.js (hunk 3 above)
Comment (robhogan) [reply examples/minimal-expo/metro.config.js:28]: Setting `nodeModulesPaths` shouldn't be necessary at all afaik?

File: docs/src/pages/api-reference/expo.mdx [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+import type { UploadRouter } from "~/app/api/uploadthing+api";
+
+export const { useImageUploader, useDocumentUploader } =
+  generateReactNativeHelpers<UploadRouter>({
+    /**
+     * Your server url.
+     * @default process.env.EXPO_PUBLIC_SERVER_URL
+     * @remarks In dev we will also try to use Expo.debuggerHost
+     */
+    url: "https://my-server.com",
```
Comment (juliusmarminge) [inline docs/src/pages/api-reference/expo.mdx:21]: Is there any way to "hook" into this variable @EvanBacon?

CleanShot 2024-04-18 at 12 06 07

**Comment by k2on** [comment]
React native for web, the File response from image picker, does not have a `size` property, which is required for startUpload

**Comment by juliusmarminge** [comment]
> I suggest this should work with RN for Web. I can help with supporting that, there are some things that I don't think work like the File response from image picker, does not have a `size` property, which is required for startUpload

What??? I would have imagined web being the place where you could expect them to follow standards the most.... Another evidence RNW is a mess

File: examples/minimal-expo/app/f/[key].tsx [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+        console.log("UPLOAD loaded", xhr.response);
+      });
+
+      xhr.open("POST", apiUrl, true);
+      xhr.setRequestHeader("Content-Length", String(blob.size));
+      xhr.setRequestHeader("Content-Type", "multipart/form-data");
+
+      xhr.send(formData);
+    })();
+  }, []);
```
Comment (juliusmarminge) [inline examples/minimal-expo/app/f/[key].tsx:70]: note to self - remove all this junk before merge

**Comment by juliusmarminge** [comment]
Pushing out another round of canary. You shouldn't have both `@uploadthing/react` and `@uploadthing/expo`, just the Expo one. (If you do end up having both make sure it's the same version used by the expo lib)

**Comment by juliusmarminge** [comment]
@younes200 I improved the URL resolving a bit to work more ootb on web

File: docs/src/pages/getting-started/expo.mdx [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+const styles = StyleSheet.create({
+  button: { ... },
+});
+```
+
+</Steps>
+
+## Deployment
+
+TODO
```
Comment (t3dotgg) [inline docs/src/pages/getting-started/expo.mdx:200]:
```suggestion
Follow the [Expo Router API Routes Deployment Guide](https://docs.expo.dev/router/reference/api-routes/#deployment)
```

Comment (t3dotgg) [reply docs/src/pages/getting-started/expo.mdx:200]: In the future we can include info on hosting server elsewhere with like Express but I think that's too nuanced for a "getting started" page

**Review by t3dotgg — APPROVED** [review]
